README.md
# Container Linux Config Transpiler

The Config Transpiler ("ct" for short) is the utility responsible for transforming a human-friendly Container Linux Config into a JSON file. This resulting file can be provided to a Container Linux machine when it first boots to provision the machine.

## Documentation

If you're looking to begin writing configs for your Container Linux machines, check out the [getting started][get-started] documentation.

The [configuration][config] documentation is a comprehensive resource specifying what options can be in a Container Linux Config.

For a more in-depth view of ct and why it exists, take a look at the [Overview][overview] document.

Please use the [bug tracker][issues] to report bugs.

[ignition]: https://github.com/coreos/ignition
[issues]: https://issues.coreos.com
[overview]: doc/overview.md
[get-started]: doc/getting-started.md
[config]: doc/configuration.md

## Examples

There are plenty of small, self-contained examples [in the documentation][examples].

[examples]: doc/examples.md

## Installation

### macOS homebrew

```bash
brew install coreos-ct
```

### Prebuilt binaries

The easiest way to get started using ct is to download one of the binaries from the [releases page on GitHub][releases].

One can use the following script to download and verify the signature of Config Transpiler:

```bash
# Specify Config Transpiler version
CT_VER=v0.6.1

# Specify Architecture
# ARCH=aarch64 # ARM's 64-bit architecture
ARCH=x86_64

# Specify OS
# OS=apple-darwin # MacOS
# OS=pc-windows-gnu.exe # Windows
OS=unknown-linux-gnu # Linux

# Specify download URL
DOWNLOAD_URL=https://github.com/coreos/container-linux-config-transpiler/releases/download

# Remove previous downloads
rm -f /tmp/ct-${CT_VER}-${ARCH}-${OS} /tmp/ct-${CT_VER}-${ARCH}-${OS}.asc /tmp/coreos-app-signing-pubkey.gpg

# Download Config Transpiler binary
curl -L ${DOWNLOAD_URL}/${CT_VER}/ct-${CT_VER}-${ARCH}-${OS} -o /tmp/ct-${CT_VER}-${ARCH}-${OS}
chmod u+x /tmp/ct-${CT_VER}-${ARCH}-${OS}

# Download and import CoreOS application signing GPG key
curl https://coreos.com/dist/pubkeys/app-signing-pubkey.gpg -o /tmp/coreos-app-signing-pubkey.gpg
gpg2 --import --keyid-format LONG /tmp/coreos-app-signing-pubkey.gpg

# Download and import CoreOS application signing GPG key if it has not already been imported
curl -L ${DOWNLOAD_URL}/${CT_VER}/ct-${CT_VER}-${ARCH}-${OS}.asc -o /tmp/ct-${CT_VER}-${ARCH}-${OS}.asc
gpg2 --verify /tmp/ct-${CT_VER}-${ARCH}-${OS}.asc /tmp/ct-${CT_VER}-${ARCH}-${OS}
```

[releases]: https://github.com/coreos/container-linux-config-transpiler/releases

### Building from source

To build from source you'll need to have the go compiler installed on your system.

```shell
git clone --branch v0.8.0 https://github.com/coreos/container-linux-config-transpiler
cd container-linux-config-transpiler
make
```

The `ct` binary will be placed in `./bin/`.

Note: Review releases for new branch versions.

## Related projects

- [https://github.com/coreos/ignition](https://github.com/coreos/ignition)
- [https://github.com/coreos/coreos-metadata/](https://github.com/coreos/coreos-metadata/)
- [https://github.com/coreos/matchbox](https://github.com/coreos/matchbox)


NEWS
18-May-2018 CONFIG-TRANSPILER v0.9.0

  Fixes:

    - Fix Makefile dependencies to ensure modified files are rebuilt
    - Do not ignore fields when explicitly set to go zero values for the etcd,
      flannel, and locksmith sections

10-Apr-2018 CONFIG-TRANSPILER v0.8.0

  Changes

    - Generate Ignition 2.2.0 configs
    - Add fields present in Ignition 2.2.0 spec
    - Emit a warning when a unit is enabled but no "Install" section is defined

  Fixes

    - Emit an error when more than one source is specified for a file

14-Feb-2018 CONFIG-TRANSPILER v0.7.0

  Features:

    - Add support for custom metadata providers
    - Add support for etcd 3.3.x
    - Automatically build docker images as part of CI

19-Jan-2018 CONFIG-TRANSPILER v0.6.1

  Fixes

    - Signatures for the v0.6.0 release binaries were generated incorrectly

12-Jan-2018 CONFIG-TRANSPILER v0.6.0

  Changes

    - Releases now include arm64 binaries
    - Enable the coreos-metadata-sshkeys service when generating OpenStack
      configs
    - Unset file modes now default to 0644 for files and 0755 for directories

  Features

    - Local files on disk can now be referenced in CLCs in place of including a
      file inline

  Fixes

    - Report URL parse errors in remote file URLs
    - Stop creating an empty file when invoked with the -out-file flag and
      conversion fails

25-Sep-2017 CONFIG-TRANSPILER v0.5.0

  Changes

    - Support Ignition spec 2.1.0

  Features

    - Common partition type GUIDs can be specified via aliases

  Fixes

    - Fix regression in validation logic causing inaccurate line and column
      reporting
    - Fix issue where locksmith fields would generate double quoted strings
      in /etc/coreos/update.conf

08-Aug-2017 CONFIG-TRANSPILER v0.4.2

  Fixes

    - Emit an error when an unsupported platform is used
    - Always emit an empty create object for non-core users to ensure Ignition
      creates them

14-Jul-2017 CONFIG-TRANSPILER v0.4.1

  Fixes

    - Fixed the build process to properly set version number

12-Jul-2017 CONFIG-TRANSPILER v0.4.0

  Features

    - Added support for vagrant-virtualbox provider
    - Added 'network_config' section for flannel
    - Added new etcd-v3.2 flags

  Fixes

    - Handle omitted versions for flannel correctly
    - Fixed envvar name LOCKSMITHD_ENDPOINT

04-May-2017 CONFIG-TRANSPILER v0.3.1

  Features

    - Added 'enable-pprof' flag in etcd

  Fixes

    - Non-string values in the etcd and flannel sections are now formatted
      correctly
    - On GCE the PUBLIC_IPV4 and PRIVATE_IPV4 dynamic data fields produced the
      incorrect environment variable names, this has been fixed

20-Apr-2017 CONFIG-TRANSPILER v0.3.0

  Features
    
    - The locksmith section gained many more fields for configuration, including
      reboot windows and etcd configuration
  
  Fixes

    - The openstack-metadata provider will now override the platform in
      coreos-metadata to set the platform to use the metadata service. This is
      necessary for the coreos-metadata unit to run successfully.

03-Apr-2017 CONFIG-TRANSPILER v0.2.2

  Features
    
    - Version field no longer required for etcd and flannel sections

28-Mar-2017 CONFIG-TRANSPILER v0.2.1

  Features

    - Added platform support for openstack-metadata

  Fixes

    - reboot-strategy under the locksmith section renamed to reboot_strategy to
      be consistent with the rest of the schema

20-Mar-2017 CONFIG-TRANSPILER v0.2.0

  Features

    - New etcd and flannel sections have been added, allowing for easy
      configuration of the service.
    - New -strict flag, that causes warnings to be treated as errors
    - New update section added, for easy selection of the OS update group and
      server
    - Dynamic data support added, allowing things like the machine's ipv4
      address to be referenced in the etcd and flannel sections
    - New docker section added, for easy configuration of flags on the docker
      daemon
    - New locksmith section added for configuring of reboot strategy

  Fixes

    - -in-file and -out-file flags were broken

29-Dec-2016 CONFIG-TRANSPILER v0.1.0

  Features

    - Support for Ignition spec v2.0.0


config/types/files.go
// Copyright 2016 CoreOS, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package types

import (
	"errors"
	"flag"
	"fmt"
	"io/ioutil"
	"net/url"
	"path"

	"github.com/coreos/container-linux-config-transpiler/config/astyaml"
	"github.com/coreos/container-linux-config-transpiler/internal/util"

	ignTypes "github.com/coreos/ignition/config/v2_1/types"
	"github.com/coreos/ignition/config/validate/astnode"
	"github.com/coreos/ignition/config/validate/report"
	"github.com/vincent-petithory/dataurl"
)

var (
	DefaultFileMode = 0644
	DefaultDirMode  = 0755

	WarningUnsetFileMode = fmt.Errorf("mode unspecified for file, defaulting to %#o", DefaultFileMode)
	WarningUnsetDirMode  = fmt.Errorf("mode unspecified for directory, defaulting to %#o", DefaultDirMode)

	ErrTooManyFileSources = errors.New("only one of the following can be set: local, inline, remote.url")
)

type FileUser struct {
	Id   *int   `yaml:"id"`
	Name string `yaml:"name"`
}

type FileGroup struct {
	Id   *int   `yaml:"id"`
	Name string `yaml:"name"`
}

type File struct {
	Filesystem string       `yaml:"filesystem"`
	Path       string       `yaml:"path"`
	User       FileUser     `yaml:"user"`
	Group      FileGroup    `yaml:"group"`
	Mode       *int         `yaml:"mode"`
	Contents   FileContents `yaml:"contents"`
}

type FileContents struct {
	Remote Remote `yaml:"remote"`
	Inline string `yaml:"inline"`
	Local  string `yaml:"local"`
}

type Remote struct {
	Url          string       `yaml:"url"`
	Compression  string       `yaml:"compression"`
	Verification Verification `yaml:"verification"`
}

type Directory struct {
	Filesystem string    `yaml:"filesystem"`
	Path       string    `yaml:"path"`
	User       FileUser  `yaml:"user"`
	Group      FileGroup `yaml:"group"`
	Mode       *int      `yaml:"mode"`
}

type Link struct {
	Filesystem string    `yaml:"filesystem"`
	Path       string    `yaml:"path"`
	User       FileUser  `yaml:"user"`
	Group      FileGroup `yaml:"group"`
	Hard       bool      `yaml:"hard"`
	Target     string    `yaml:"target"`
}

func (f File) ValidateMode() report.Report {
	if f.Mode == nil {
		return report.ReportFromError(WarningUnsetFileMode, report.EntryWarning)
	}
	return report.Report{}
}

func (d Directory) ValidateMode() report.Report {
	if d.Mode == nil {
		return report.ReportFromError(WarningUnsetDirMode, report.EntryWarning)
	}
	return report.Report{}
}

func (fc FileContents) Validate() report.Report {
	i := 0
	if fc.Remote.Url != "" {
		i++
	}
	if fc.Inline != "" {
		i++
	}
	if fc.Local != "" {
		i++
	}
	if i > 1 {
		return report.ReportFromError(ErrTooManyFileSources, report.EntryError)
	}
	return report.Report{}
}

func init() {
	register(func(in Config, ast astnode.AstNode, out ignTypes.Config, platform string) (ignTypes.Config, report.Report, astnode.AstNode) {
		r := report.Report{}
		files_node, _ := getNodeChildPath(ast, "storage", "files")
		for i, file := range in.Storage.Files {
			if file.Mode == nil {
				file.Mode = util.IntToPtr(DefaultFileMode)
			}
			file_node, _ := getNodeChild(files_node, i)
			newFile := ignTypes.File{
				Node: ignTypes.Node{
					Filesystem: file.Filesystem,
					Path:       file.Path,
					User: ignTypes.NodeUser{
						ID:   file.User.Id,
						Name: file.User.Name,
					},
					Group: ignTypes.NodeGroup{
						ID:   file.Group.Id,
						Name: file.Group.Name,
					},
				},
				FileEmbedded1: ignTypes.FileEmbedded1{
					Mode: *file.Mode,
				},
			}

			if file.Contents.Inline != "" {
				newFile.Contents = ignTypes.FileContents{
					Source: (&url.URL{
						Scheme: "data",
						Opaque: "," + dataurl.EscapeString(file.Contents.Inline),
					}).String(),
				}
			}

			if file.Contents.Local != "" {
				// The provided local file path is relative to the value of the
				// --files-dir flag.
				filesDir := flag.Lookup("files-dir")
				if filesDir == nil || filesDir.Value.String() == "" {
					err := errors.New("local files require setting the --files-dir flag to the directory that contains the file")
					flagReport := report.ReportFromError(err, report.EntryError)
					if n, err := getNodeChildPath(file_node, "contents", "local"); err == nil {
						line, col, _ := n.ValueLineCol(nil)
						flagReport.AddPosition(line, col, "")
					}
					r.Merge(flagReport)
					continue
				}
				localPath := path.Join(filesDir.Value.String(), file.Contents.Local)
				contents, err := ioutil.ReadFile(localPath)
				if err != nil {
					// If the file could not be read, record error and continue.
					convertReport := report.ReportFromError(err, report.EntryError)
					if n, err := getNodeChildPath(file_node, "contents", "local"); err == nil {
						line, col, _ := n.ValueLineCol(nil)
						convertReport.AddPosition(line, col, "")
					}
					r.Merge(convertReport)
					continue
				}

				// Include the contents of the local file as if it were provided inline.
				newFile.Contents = ignTypes.FileContents{
					Source: (&url.URL{
						Scheme: "data",
						Opaque: "," + dataurl.Escape(contents),
					}).String(),
				}
			}

			if file.Contents.Remote.Url != "" {
				source, err := url.Parse(file.Contents.Remote.Url)
				if err != nil {
					// if invalid, record error and continue
					convertReport := report.ReportFromError(err, report.EntryError)
					if n, err := getNodeChildPath(file_node, "contents", "remote", "url"); err == nil {
						line, col, _ := n.ValueLineCol(nil)
						convertReport.AddPosition(line, col, "")
					}
					r.Merge(convertReport)
					continue
				}

				// patch the yaml tree to look like the ignition tree by making contents
				// the remote section and changing the name from url -> source
				asYamlNode, ok := file_node.(astyaml.YamlNode)
				if ok {
					newContents, _ := getNodeChildPath(file_node, "contents", "remote")
					newContentsAsYaml := newContents.(astyaml.YamlNode)
					asYamlNode.ChangeKey("contents", "contents", newContentsAsYaml)

					url, _ := getNodeChild(newContents.(astyaml.YamlNode), "url")
					newContentsAsYaml.ChangeKey("url", "source", url.(astyaml.YamlNode))
				}

				newFile.Contents = ignTypes.FileContents{Source: source.String()}

			}

			if newFile.Contents == (ignTypes.FileContents{}) {
				newFile.Contents = ignTypes.FileContents{
					Source: "data:,",
				}
			}

			newFile.Contents.Compression = file.Contents.Remote.Compression
			newFile.Contents.Verification = convertVerification(file.Contents.Remote.Verification)

			out.Storage.Files = append(out.Storage.Files, newFile)
		}
		for _, dir := range in.Storage.Directories {
			if dir.Mode == nil {
				dir.Mode = util.IntToPtr(DefaultDirMode)
			}
			out.Storage.Directories = append(out.Storage.Directories, ignTypes.Directory{
				Node: ignTypes.Node{
					Filesystem: dir.Filesystem,
					Path:       dir.Path,
					User: ignTypes.NodeUser{
						ID:   dir.User.Id,
						Name: dir.User.Name,
					},
					Group: ignTypes.NodeGroup{
						ID:   dir.Group.Id,
						Name: dir.Group.Name,
					},
				},
				DirectoryEmbedded1: ignTypes.DirectoryEmbedded1{
					Mode: *dir.Mode,
				},
			})
		}
		for _, link := range in.Storage.Links {
			out.Storage.Links = append(out.Storage.Links, ignTypes.Link{
				Node: ignTypes.Node{
					Filesystem: link.Filesystem,
					Path:       link.Path,
					User: ignTypes.NodeUser{
						ID:   link.User.Id,
						Name: link.User.Name,
					},
					Group: ignTypes.NodeGroup{
						ID:   link.Group.Id,
						Name: link.Group.Name,
					},
				},
				LinkEmbedded1: ignTypes.LinkEmbedded1{
					Hard:   link.Hard,
					Target: link.Target,
				},
			})
		}
		return out, r, ast
	})
}
